### PR TITLE
template min.css and min.js etc

### PR DESCRIPTION
**Pull request by brianteeman**
It is currently possible to upload a file.min.css, file.min.js, file.asset.json

However it is not possible to rename or create a new file as we had a check to prevent having a . in the name

There really is no need for this restriction. It dates back 10 years to a security fix but it wasn't needed as the rest of the security fix is still valid.

NOTE: This does not add .min.css or .min.js or asset.json to the list of filetypes. They are still just css, js and json but now you can put a dot in the filename

image

image

### CODE REVIEW

**Comment by N6REJ** [comment]
I have tested this item :white_check_mark: successfully on 1e62228bd7816a8a38904d7f664b9b8fedb4e4fc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by Kostelano** [comment]
I have tested this item :white_check_mark: successfully on 1e62228bd7816a8a38904d7f664b9b8fedb4e4fc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by richard67** [comment]
@brianteeman Does this PR solve the 2 issues #38455 and #30908 so they can be closed?

**Comment by bembelimen** [comment]
Just from code review: we have to be careful to not allow: "...hello...bla.min"

**Comment by brianteeman** [comment]
@bembelimen I did wonder about that but I assumed (maybe incorrectly) that as we didnt allow the slash it would not be a problem?

**Comment by brianteeman** [comment]
please remove the RTC on this.

@bembelimen was correct and we need to prevent a .. in the filename. I was only thinking of security but actually it blows away the php if its allowed. - will update this PR later

**Comment by richard67** [comment]
Back to pending due to reason stated in previous comment.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by brianteeman** [comment]
I really dont know what I was thinking when I wrote this - its utter garbage

**Comment by brianteeman** [comment]
1. It must also prevent ..
2. The dot should have been escaped with a \
3. My reading of the code says that if the file has a character not alphanumeric or dash or underscore and now dot then you cannot create a new file of that name. But thats not what is happening. 

Even before this PR if you try to create a file with a dot in it then it refuses to create the file which is expected.
However if you try to create a file with another non matching character such as brian&broken I would expect that to fail as well. It doesnt. it will be created as brianbroken

**Comment by brianteeman** [comment]
Please retest and confirm that you can now create, rename files with a dot in them AND that you cannot when there are two or more dots

File: administrator/components/com_templates/src/Controller/TemplateController.php [hunk 1]
```diff
@@ -668,7 +668,7 @@ public function renameFile()
             $this->setMessage(Text::_('COM_TEMPLATES_ERROR_RENAME_ASSET_FILE'), 'warning');
             $url = 'index.php?option=com_templates&view=template&id=' . $id . '&file=' . $file . '&isMedia=' . $isMedia;
             $this->setRedirect(Route::_($url, false));
-        } elseif (!preg_match('/^[a-zA-Z0-9-_]+$/', $newName)) {
+        } elseif (!preg_match('/^[a-zA-Z0-9-_.]+$/', $newName) || strpos($newName, '..') !== false) {
```
Comment (richard67) [inline administrator/components/com_templates/src/Controller/TemplateController.php]: In the previous change the dot is escaped, but here and in the changes below it is not. That seems strange to me.

Comment (brianteeman) [reply administrator/components/com_templates/src/Controller/TemplateController.php]: thanks

**Comment by zero-24** [comment]
Whats with a "invisible space" IIRC thats not catched by strpos. Or some kind of none breaking change as well as some unicode char thats confusing the filename/filesystem? didnt we have a filter class where we could throw this file name against too?

**Comment by brianteeman** [comment]
@zero-24 The file is already going through a filter before this iirc. If not then its beyond the scope of this PR as it would be possible before this PR as well. please test those yourself.

**Comment by Kostelano** [comment]
I have tested this item :white_check_mark: successfully on fa2068eaa65eb69fba1f7cc2606934429793d0a4<br /><br />Tested again. With one dot - the file was created, with two or more - an error.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on fa2068eaa65eb69fba1f7cc2606934429793d0a4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by jwaisner** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by bembelimen** [comment]
"." should not be allowed at the beginning of the file (as then the file is hidden) and also not at the end of the file (to avoid again the ".." with the extension.

**Comment by brianteeman** [comment]
>  should not be allowed at the beginning of the name
Its not - either before or after this pr. In both cases the leading dot is stripped

>  also not at the end of the name
yeah thats a bug in the regex will need to fix thaat after the holiday

**Comment by brianteeman** [comment]
please remove rtc for now

**Comment by richard67** [comment]
Back to pending.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by brianteeman** [comment]
Sorry I must have got distracted. I had the correct regex already written but for some reason I didnt include it.

You can test the regex here https://rextester.com/RULEP70144

**Comment by Kostelano** [comment]
I have tested this item :white_check_mark: successfully on e291abbb929f91beaaa51844d790b23259134efa<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on e291abbb929f91beaaa51844d790b23259134efa<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by N6REJ** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by N6REJ** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by fancyFranci** [comment]
I would really like to merge it, but regex with this size are tricky and need more tests. I'm wondering where the underscore in the first regex went? But 4.3 has the time to test intensively, so I'm moving the PR.

**Comment by brianteeman** [comment]
There will be zero extra tests. Look at who has already tested it.

**Comment by obuisard** [comment]
As @fancyFranci noticed, it does not allow underscore in the regexp anymore. A file name like joomla_78.min.css will fail

**Comment by bembelimen** [comment]
Sorry, don't want to be the party pooper again. So I tested it:

## Before:

### Edit/Create file:
- _bla.scss => works
- bla.min.css => does not work
- text-file.php => works
- .hui.txt => converted to "hui.txt" andsaved (should probably give an error)

## After:

### Create file:
- _bla.scss => does not work
- bla.min.css => works
- text-file.php => does not work
- .hui.txt => converted to "hui.txt"

### Edit file:
- _bla.scss => works
- bla.min.css => does not work
- text-file.php => works
- .hui.txt => converted to "hui.txt" and saved

I also think I was wrong regaring the "." at the beginning, there is no test needed as the getCmd filter already removed the ".". So I think ```(?!.*\\..*\\..*)``` is not needed at the beginning?

**Comment by bembelimen** [comment]
Back to pending<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38458">issues.joomla.org/tracker/joomla-cms/38458</a>.</sub>

**Comment by brianteeman** [comment]
Please make up your minds.

Closed

**Comment by obuisard** [comment]
The difference in create and edit comes from the fact that they both use a different regular expression, therefore you get different results. Nothing major :-)

**Comment by obuisard** [comment]
@brianteeman this is a good PR, it addresses issues encountered by users.
Would you mind re-opening it? Thank you.

**Comment by richard67** [comment]
@obuisard There is PR #38987 .

**Pull request closed without merging**